cl/disclosures/tasks.py

```diff
@@ -29,3 +42,429 @@ def make_financial_disclosure_thumbnail_from_pdf(pk: int) -> None:
+            ]["rows"]
+        ]
+    )
+
+    SpouseIncome.objects.bulk_create(
+        [
+            SpouseIncome(
+                financial_disclosure=disclosure,
+                redacted=any(v["is_redacted"] for v in spouse_income.values()),
+                date_raw=spouse_income["Date"]["text"],
+                source_type=spouse_income["Source and Type"]["text"],
+            )
+            for spouse_income in extracted_data["sections"][
+                "Non Investment Income Spouse"
+            ]["rows"]
+        ]
+    )
+
+
+def get_aws_url(data: Dict[str, Union[str, int, list]]) -> str:
+    """Get URL saved to download filepath
+
+    :param data: File data
+    :return: URL or first URL on AWS
+    """
+    if data["disclosure_type"] == "jw" or data["disclosure_type"] == "single":
+        url = data["url"]
+    else:
+        url = data["urls"][0]
+    return url
+
+
+def get_disclosure_from_pdf_path(disclosure_url: str):
+    """Convenience method to get disclosure from download filepath
+
+    :param disclosure_url: The URL of the first link (if there are more than
+    one) of the source FD tiff(s)/PDF
+    :return: Financial Disclosure object
+    """
+    from cl.disclosures.models import FinancialDisclosure
+
+    return FinancialDisclosure.objects.get(download_filepath=disclosure_url)
+
+
+def has_been_pdfed(disclosure_url: str) -> Optional[str]:
+    """Has file been PDFd from tiff and saved to AWS.
+
+    :param disclosure_url: The URL of the first link (if there are more than
+    one) of the source FD tiff(s)/PDF
+    :return: Path to document or None
+    """
+    from cl.disclosures.models import FinancialDisclosure
+
+    disclosures = FinancialDisclosure.objects.filter(
+        download_filepath=disclosure_url
+    )
+    if disclosures.exists():
+        return (
+            f"https://{settings.AWS_S3_CUSTOM_DOMAIN}/"
+            f"{disclosures[0].filepath}"
+        )
+
+
+def generate_or_download_disclosure_as_pdf(
+    data: Dict[str, Union[str, int, List[str]]],
+    pdf_url: Optional[str],
+) -> requests.Response:
+    """Generate or download PDF content from images or urls.
+
+    :param data: Data to process.
+    :param pdf_url: The URL of PDF in S3
+    :return: Response containing PDF
+    """
+    if pdf_url:
+        logger.info(f"Downloading PDF: {pdf_url}")
+        return requests.get(pdf_url, timeout=60 * 20)
+    elif data["disclosure_type"] == "jw":
+        logger.info(f"Downloading JW PDF: {quote(data['url'], safe=':/')}")
+        return requests.get(data["url"], timeout=60 * 20)
+    elif data["disclosure_type"] == "single":
+        urls = [data["url"]]
+    else:
+        urls = data["urls"]
+    logger.info(f"Processing url:{quote(urls[0], safe=':/')}")
+    return requests.post(
+        settings.BTE_URLS["images-to-pdf"]["url"],
+        json=json.dumps({"urls": urls}),
+        timeout=settings.BTE_URLS["images-to-pdf"]["timeout"],
+    )
+
+
+@app.task
```

Comment:
Why not add some retries with exponential backoff here?

Reply:
I was concerned that - if we get into a trump situation that times out after two hours we dont want to retry and retry and waste lots of valuable time

Reply:
Hm, yeah, it'd be nice not to have to do that. You can be pretty granular with your retries though. For example, you can configure the task to have n retries, and then you can only catch the particular ones you want to retry. Check out some examples in the recap tasks? (grep for "retry")

Reply:
√. added.

Reply:
I figured - its actually not an issue- a timeout should already be handled in the code itself and should continue processing.  Its more the weird connection timeout that was more concerning that I think should be handled.